### ISSUE TITLE
[Bug] bufr2ioda fails with multiple subsets in a bufr file (e.g. NSST)

### ISSUE THREAD

**Issue by ShastriPaturi** [JCSDA-internal/ioda-converters#1139]
## Current behavior (describe the bug)
>bufr2ioda fails with multiple subsets in a bufr file (e.g. NSST)

An example of the yaml is attached:

```yaml
observations:
  - obs space:
      name: bufr
      obsdatain: "./gdas.t06z.nsstbufr"

      exports:
        #group_by_variable: depth  # Optional
        variables:
          timestamp:
            datetime:
              year: "*/YEAR[1]"
              month: "*/MNTH[1]"
              day: "*/DAYS[1]"
              hour: "*/HOUR[1]"
              minute: "*/MINU[1]"

          rcptimestamp:
            datetime:
              year: "*/RCPTIM/RCYR"
              month: "*/RCPTIM/RCMO"
              day: "*/RCPTIM/RCDY"
              hour: "*/RCPTIM/RCHR"
              minute: "*/RCPTIM/RCMI"

          longitude:
            query: "*/CLON"
          latitude:
            query: "*/CLAT"
```

The following error is generated when running bufr2ioda:
PRE-MAIN-WARNING Warning: Query String */RCPTIM/RCYR didn't apply to subset NC031002
PRE-MAIN-WARNING Warning: Query String */CLON didn't apply to subset NC031002
PRE-MAIN-WARNING Warning: Query String */RCPTIM/RCDY didn't apply to subset NC031002
PRE-MAIN-WARNING Warning: Query String */RCPTIM/RCMI didn't apply to subset NC031002
PRE-MAIN-WARNING Warning: Query String */SST1 didn't apply to subset NC031002
PRE-MAIN-WARNING Warning: Query String */RCPTIM/RCHR didn't apply to subset NC031002
PRE-MAIN-WARNING Warning: Query String */YEAR[1] didn't apply to subset NC031002
PRE-MAIN-WARNING Warning: Query String */DAYS[1] didn't apply to subset NC031002
**terminate called after throwing an instance of 'std::length_error'
  what():  vector::_M_default_append**

**Comment by ShastriPaturi** [comment]
@rmclaren I concatenated two bufr files and was able to run the converter. 
I noticed that for the NSST bufr file, the print_queries.x dump has empty subset in the end Shown below), is that an issue:

NC031002
 Dimensioning Sub-paths:

 Queries:

**Comment by rmclaren** [comment]
HERA is down right now, so I'm unable to see an example of this data set. Could you run the gettab utility on this bufr file and post the result to this thread?

**Comment by rmclaren** [comment]
Do things change when you comment out the `rcptimestamp` section in your YAML? The answer to your question is probably...

**Comment by ShastriPaturi** [comment]
No change when I comment out NSST_queries.txt in the YAML when I use the NSST bufr (gdas.t06z.nsstbufr) that has no info for NC031002 subset.

**Comment by ShastriPaturi** [comment]
> gettab

Where can I find the gettab utility, you mentioned?

**Comment by rmclaren** [comment]
That one is part of NCEPLIB-bufr... So <your install directory>/bin/gettab

**Comment by rmclaren** [comment]
Would it be possible to provide a sample of the BUFR data and YAML file. Unfortunately I don't have access to Orion right now... I do have access to HERA however.

**Comment by ShastriPaturi** [comment]
@rmclaren I have put the BUFR data and the yaml I used on hera here:
/scratch2/NCEPDEV/marine/Shastri.Paturi/forRon

**Comment by rmclaren** [comment]
@ShastriPaturi unfortunately the permissions set on the bufr file make it impossible for me to look at it...

**Comment by ShastriPaturi** [comment]
@rmclaren apologies.
I changed the permission.

**Comment by rmclaren** [comment]
So this BUFR file seems to have the weird property that  there are messages for subset NC031002 without any actual subset data to go along with them (they are empty). So in the standard loop:

```
    while (ireadmg_f(fileUnit, current_subset, &iddate, SubsetLen) == 0)
    {
        auto msg_subset = std::string(current_subset);
        msg_subset.erase(
                remove_if(msg_subset.begin(), msg_subset.end(), isspace), msg_subset.end());

        status_f(fileUnit, &bufrLoc, &il, &im);
        dataProvider.updateData(bufrLoc);

        if (msg_subset == subset)
        {
            while (ireadsb_f(fileUnit) == 0)
            {```

when `msg_subset == NC031002` then `ireadsb` never suceeds...

**Comment by ShastriPaturi** [comment]
That is what we discussed @rmclaren last Thursday. Thanks for confirming.

**Comment by rmclaren** [comment]
@ShastriPaturi I just wanted to look into it a little more... Not sure how to handle this at the moment.

**Comment by ShastriPaturi** [comment]
FYI: @rmclaren, I have made @ilianagenkova be aware of this issue.

**Comment by ilianagenkova** [comment]
@rmclaren and @ShastriPaturi,  You may be aware that *nsstbufr is nothing else but collection of data present in other bufr dump files. In this case, 031/002 (a.k.a. NC031002, b031/xx002) is _tesac_ type data which exists as  stand alone file:
 _binv  /lfs/h1/ops/prod/com/obsproc/v1.1/gdas.20230202/00/atmos/gdas.t00z.tesac.tm00.bufr_d
type        messages       subsets         bytes 
NC031002         788          **2128**        439919        2.70
TOTAL            788          2128        439919_

but also in nsstbufr:
 _binv  /lfs/h1/ops/prod/com/obsproc/v1.1/gdas.20230202/00/atmos/gdas.t00z.nsstbufr 
 
type        messages       subsets         bytes
 
NC001001         113           142         35492        1.26
NC001013         393           521        120356        1.33
NC001002         305           311        132226        1.02
NC001003         719           783        186016        1.09
NC001004          88           528         95380        6.00
NC001007          43            50          8722        1.16
NC001102        6400          7113       1575332        1.11
NC001103        2000          5419        893844        2.71
NC001101         631           723        177074        1.15
NC001113        3462          6537       1464134        1.89
NC001104        3349         25634       4661654        7.65
NC031002         788          **2128**        439919        2.70
NC031001          11             9         58548        0.82
NC031003          30           734         86747       24.47
NC031005         172           296       1114368        1.72
TOTAL          18504         50928      11049812_

In the example above there is 2128 TESAC obs (actually TESAC is the format name for reporting operational oceanographic data).  If there is no TESAC data, the *tesac*.bufr_d would most probably be 0 size and you won't fine NC031002 in nsstbufr.

There are a few observation types that may not be present from time to time. Could the converter code check for that?  
What else can I help with?

**Comment by ShastriPaturi** [comment]
Thanks @ilianagenkova. This is sufficient for now.

**Comment by rmclaren** [comment]
Been looking for a quick fix, unfortunately I keep running into different issues...

**Comment by rmclaren** [comment]
It turns out that the empty subsets didn't have anything to do with the problem. The root cause was that the NCEP meta data table data (at the beginning of the file) isn't consistent with the data tables in the rest of the file for some of the subsets (ex: NC031002). I updated a version of the code in the following branch **feature/query_bad_ncep_table** which can handle the situation... 

I need a version of that **gdas.t06z.nsstbufr** to generate a unit test. It should be much smaller (no more than 200KB) and preserve maybe 10 samples of each subset type in the original file (needs to be able to demonstrate same problem as the original data file with the original code)...

**Comment by rmclaren** [comment]
Once I have the unit test file I can create the test and submit a pull request.

**Comment by ilianagenkova** [comment]
@rmclaren We could i) split_by_subset the gdas.t06z.nsstbufr , ii) "clip" each subset, ii) and concatenate back to one nsstbufr. Do you want any of the subsets to be missing?  
@ShastriPaturi I will show you how to do that

**Comment by rmclaren** [comment]
@ilianagenkova Given the branch I referenced above *feature/query_bad_ncep_table* it should not be necessary to split the subsets, or anything like that. Empty messages (messages with no subsets) are fine and should not pose an issue. Like I said the root problem is that the meta data in these files is messed up. The branch takes care of this, however, and can substitute the table data it finds in the message itself if there is a problem with the master table for the BUFR file, so it can now read all the data in these files. 

In short just use *feature/query_bad_ncep_table*. As soon as we make an appropriate  test file we can make a PR and merge it into the develop branch.

**Comment by ShastriPaturi** [comment]
> @rmclaren We could i) split_by_subset the gdas.t06z.nsstbufr , ii) "clip" each subset, ii) and concatenate back to one nsstbufr. Do you want any of the subsets to be missing? @ShastriPaturi I will show you how to do that

Thanks @ilianagenkova.

**Comment by ShastriPaturi** [comment]
> @ilianagenkova Given the branch I referenced above **feature/query_bad_ncep_table** it should not be necessary to split the subsets, or anything like that. Empty messages (messages with no subsets) are fine and should not pose an issue. Like I said the root problem is that the meta data in these files is messed up. The branch takes care of this, however, and can substitute the table data it finds in the message itself if there is a problem with the master table for the BUFR file, so it can now read all the data in these files.
> 
> In short just use **feature/query_bad_ncep_table**. As soon as we make an appropriate test file we can make a PR and merge it into the develop branch.

Thanks @rmclaren. I will test the gdas.t06z.nsstbufr that I have and have tested earlier and let you know.

**Comment by rmclaren** [comment]
I haven't heard anything in a while, Do you want me to create the dataset so that I can submit the PR for this?

**Comment by ShastriPaturi** [comment]
@rmclaren I just tested the **gdas.t06z.nsstbufr**  with the **feature/query_bad_ncep_table** and it works.
@ilianagenkova can you please help me with subsetting the **gdas.t06z.nsstbufr** file for @rmclaren to create the PR and push the fix?

**Comment by ilianagenkova** [comment]
@ShastriPaturi , I generated a smaller nsstbufr file (on Dogwood):
/lfs/h2/emc/obsproc/noscrub/iliana.genkova/NSST_Shastri/gdas.t00z.nsstbufr.small (for 20230313/00)

@rmclaren , you asked for ~200K size file, but the tools I use won't allow me to make a smaller sample.

I will email you to explain how the gdas.t00z.nsstbufr.small  file was produced.

**Comment by PatNichols** [comment]
@ShastriPaturi I am going to close this since the code in question is now replaced by a newer version. You can reopen it if you are still seeing problems.
